components/builder-api/src/http/handlers.rs

```diff
@@ -163,18 +172,37 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
                 do_channel_creation(req, ident.get_origin(), &channel, session_id)?;
             }
 
-            do_promotion(req, ident, &channel, session_id)?;
+            let mut project_list = origin_map.entry(ident.get_origin().to_string()).or_insert(
+                Vec::new(),
+            );
+            project_list.push(project);
         } else {
-            projects.push(project.get_ident().to_string());
+            failed_projects.push(project.get_ident().to_string());
+        }
+    }
+
+    // Now that we've sorted all the projects into a HashMap keyed on origin, we can process them,
+    // one origin at a time. We do "core" first.
+    if let Some(core_projects) = origin_map.remove("core") {
+        let promote_result = do_group_promotion(&channel, core_projects, "core");
+        if !promote_result.is_ok() {
+            return promote_result;
+        }
+    }
+
+    for (origin, projects) in origin_map.iter() {
+        let promote_result = do_group_promotion(&channel, projects.to_vec(), &origin);
+        if !promote_result.is_ok() {
+            return promote_result;
```

Comment:
The more I think about how to solve this, the messier it feels.  Right now, pretty much everywhere in our codebase that I've seen, when an operation returns an error, we immediately return that error.  The returned error bubbles up, eventually reaching the very top of the call chain in `main.rs` where the process terminates with a non-zero exit code.  This follows that same pattern.

In thinking through how this would work if we didn't do that, and instead continued on if there was an error, we'd need some way of keeping track of the errors as they occurred and storing them for later use.  Then, once we had finished trying to promote everything, we'd have to somehow return this collection of errors and then figure out how to elegantly display multiple errors to the user.  Even if we solved the problem of displaying those errors to the user in a way that made sense, how would these multiple errors interact with Rust's notion of `Result` which accounts for a single success and single error value.  What gets bubbled back up to `main.rs`?

Maybe I'm just thinking about this wrong, but going this route feels like it opens a Pandora's Box of issues.